### DISCUSSION TITLE
Should the <script> tag in HTML files not be counted as Javascript?

### DISCUSSION THREAD

**Discussion by Sainan** [github-linguist/linguist#5248] [Classification] [answered]
Consider my pointless project at https://github.com/UtopiaLang/Utopia and note that it has HTML listed, no Javascript. However, this is solely because of [this HTML file](https://github.com/UtopiaLang/Utopia/blob/senpai/example-use-lib-via-wasm.html), which exclusively features `<script>` tags, practically making it mostly javascript, if not entirely javascript.

This issue may be even worse with .php files, because they are HTML [unless in phpmode](https://www.php.net/manual/en/language.basic-syntax.phpmode.php).

**Comment by Nixinova** [comment, marked as answer]
Linguist cannot detect sub-languages inside files. Since the file is indeed an HTML file, it is classified as HTML and HTML only.

**Reply by Sainan** [reply]
Excuses, excuses.

**Reply by lildude** [reply]
As pointed out by @Nixinova Linguist does not perform multiple-language per-file analysis and this is a deliberate design decision as it would entail having to open, read and perform a line-by-line analysis of every single file uploaded to GitHub. This would be incredibly resource intensive and unreliable for very little gain. Linguist tries to get as close as possible without trying to be perfect thus it has overrides for those cases where people don't like Linguist's choice. 

This is no different to VSCode which also detects your file as HTML:

![CleanShot 2021-03-04 at 09 27 14](https://user-images.githubusercontent.com/627280/109942392-0ee51980-7ccc-11eb-9120-b294bc70302d.png)

**Reply by Sainan** [reply]
I can understand the resource costs not being worth it, but I don't think the comparison with an editor is apt; an editor can make such a broad assertion, because the PHP formatting will also contains within it HTML formatting, and the HTML formatting will contain JS formatting.

**Reply by lildude** [reply]
I was trying to highlight the fact that, as you've pointed out, the editor has the knowledge and ability to more precisely determine the language but hasn't. I don't know the intricacies of how VSCode determined the language, but it appears to be based on the extension, like Linguist. And like Linguist, you can override that choice made.

**Reply by Sainan** [reply]
I think an editor would be foolish to determine the language more precisely, because the file is the canvas and the most broad formatting with the most sub-formattings will be able to accomodate the most use cases, but files on Github are more static than that, and if the goal is to say which languages are ***actually used***, then the broad stroke is a cheap guess.

**Reply by simonafield** [reply]
To force Linguist to consider the contents of your <script> block within HTML, you can move the contents of <script> into a different file with the `.js` extension, `script.js` for this example. Then, in your HTML, you can replace the now empty <script> block with the following:

>  <script src="script.js"></script>

OR

> <script src="script.js" />
(self closing)

If you don't want Linguist to detect this JavaScript, leave it in the HTML <script> block.

**Reply by Sainan** [reply]
Nice job reviving a 3 y/o thread. Yes, that is what I do now when I want to have it display correctly on Github.

**Reply by simonafield** [reply]
Haha sorry for my lateness. I guess it's more for those who stumble on this discussion in search of the solution. (:
